### PR TITLE
[feat]: [CDS-47030]: ASG NG Canary Delete - All manager changes 

### PR DESCRIPTION
**Pull request by lovish1999**
## Describe your changes
Added Manager Side Implementation for ASG NG Swimlane Canary delete step. We take the Canary ASG Name (from sweeping output via canary deploy step) and Infra Configuration and pass them to delegate side for the deletion of Canary ASG. 

## Checklist
- [x] I've documented the changes in the PR description.
- [x] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- Immutable delegate `trigger publish-delegate`
</details>

<details>
  <summary>PR Check triggers</summary>

You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ti0, ti1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
    - CodeFormat: `trigger codeformat`
    - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- TI-bootstrap: `trigger ti0`
- TI-bootstrap1: `trigger ti1`
- TI-bootstrap2: `trigger ti2`
- TI-bootstrap3: `trigger ti3`
- TI-bootstrap4: `trigger ti4`
- FunctionalTest1: `trigger ft1`
- FunctionalTest2: `trigger ft2`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- Trigger all Checks: `trigger smartchecks`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/41429)

### CODE REVIEW

**Review by lovish1999 — COMMENTED** [review]
AsgCanaryDeleteDataOutcome & AsgPrepareRollbackDataPassThroughData classes need some refactoring. Will do that.

**Review by vitsafronovici — APPROVED** [review]
> Reviewed 45 of 45 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/41429)* status: 45 file(s) to review. 1 of 2 approvals obtained.

**Review by vitsafronovici — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/41429)* status: 45 file(s) to review. 1 change requests. 0 of 2 approvals obtained. 

___
*[`950-delegate-tasks-beans/src/main/java/io/harness/delegate/task/aws/asg/AsgCommandTypeNG.java` line 10 at r1](https://reviewable.io/reviews/harness/harness-core/41429#-NJ5cDGT-ch0j-oHRkSO:-NJ5cDGUCA0okbwcL_xh:b-pi9d9n) ([raw file](https://github.com/harness/harness-core/blob/8d015a76127ac229a718b5a5d614d42ed7962269/950-delegate-tasks-beans/src/main/java/io/harness/delegate/task/aws/asg/AsgCommandTypeNG.java#L10)):*
> ```Java
> package io.harness.delegate.task.aws.asg;
> 
> public enum AsgCommandTypeNG { ASG_CANARY_DEPLOY, ASG_CANARY_DELETE }
> ```

remove this

**Review by yogesh-chauhan — COMMENTED** [review]
> Reviewed 9 of 45 files at r1.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/41429)* status: 36 file(s) to review. 1 change requests. 0 of 2 approvals obtained.

**Comment by arvind-choudhary-h** [comment]
*[`125-cd-nextgen/src/main/java/io/harness/cdng/creator/CDCreatorUtils.java` line 50 at r3](https://reviewable.io/reviews/harness/harness-core/41429#-NJEMoUHDC8FUxnmIdKK:-NJEMoUHDC8FUxnmIdKL:b-h8302v) ([raw file](https://github.com/harness/harness-core/blob/ad8cd66cb394b17d87051427e81be66d86d2dc65/125-cd-nextgen/src/main/java/io/harness/cdng/creator/CDCreatorUtils.java#L50)):*
> ```Java
>         StepSpecTypeConstants.ELASTIGROUP_ROLLBACK, StepSpecTypeConstants.ELASTIGROUP_SETUP,
>         StepSpecTypeConstants.TERRAGRUNT_PLAN, StepSpecTypeConstants.TERRAGRUNT_APPLY,
>         StepSpecTypeConstants.TERRAGRUNT_DESTROY, StepSpecTypeConstants.TERRAGRUNT_ROLLBACK);
> ```

Let's add in sequence -- StepSpecTypeConstants.ASG_CANARY_DELETE should be at end.

**Comment by arvind-choudhary-h** [comment]
*[`950-delegate-tasks-beans/src/main/java/io/harness/delegate/task/aws/asg/AsgCanaryDeployResult.java` line 22 at r3](https://reviewable.io/reviews/harness/harness-core/41429#-NJENVZB8S1xWAtfi3D9:-NJENVZB8S1xWAtfi3DA:b-yhyd29) ([raw file](https://github.com/harness/harness-core/blob/ad8cd66cb394b17d87051427e81be66d86d2dc65/950-delegate-tasks-beans/src/main/java/io/harness/delegate/task/aws/asg/AsgCanaryDeployResult.java#L22)):*
> ```Java
> public class AsgCanaryDeployResult {
>   private String region;
>   //private List<AsgTask> asgTasks;
> ```

let's remove not required fields.

**Comment by arvind-choudhary-h** [comment]
*[`950-delegate-tasks-beans/src/main/java/io/harness/delegate/task/aws/asg/AsgCanaryDeployResult.java` line 22 at r3](https://reviewable.io/reviews/harness/harness-core/41429#-NJENVZB8S1xWAtfi3D9:-NJENZFx2knuOtnL6VVQ:b-qghwja) ([raw file](https://github.com/harness/harness-core/blob/ad8cd66cb394b17d87051427e81be66d86d2dc65/950-delegate-tasks-beans/src/main/java/io/harness/delegate/task/aws/asg/AsgCanaryDeployResult.java#L22)):*
<details><summary><i>Previously, arvind-choudhary-h (Arvind Choudhary) wrote…</i></summary><blockquote>

let's remove not required fields.
</blockquote></details>

The commented line

**Review by arvind-choudhary-h — COMMENTED** [review]
> Reviewed 45 of 45 files at r1, 20 of 37 files at r2, 27 of 27 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/41429)* status: 40 file(s) to review. 1 change requests. 0 of 2 approvals obtained.

File: 950-delegate-tasks-beans/src/main/java/io/harness/delegate/task/aws/asg/AsgInfraType.java [hunk 1]
```diff
@@ -0,0 +1,10 @@
+/*
+ * Copyright 2022 Harness Inc. All rights reserved.
+ * Use of this source code is governed by the PolyForm Free Trial 1.0.0 license
+ * that can be found in the licenses directory at the root of this repository, also available at
+ * https://polyformproject.org/wp-content/uploads/2020/05/PolyForm-Free-Trial-1.0.0.txt.
+ */
+
+package io.harness.delegate.task.aws.asg;
+
+public enum AsgInfraType { ASG }
```
Comment (vitsafronovici) [inline 950-delegate-tasks-beans/src/main/java/io/harness/delegate/task/aws/asg/AsgInfraType.java:10]: it is also used in ECS, but I don't understand it's purpose. why need enum with one value. why not to use InfrastructureKind.ASG instead

**Comment by arvind-choudhary-h** [comment]
*[`950-delegate-tasks-beans/src/main/java/software/wings/beans/TaskType.java` line 423 at r6](https://reviewable.io/reviews/harness/harness-core/41429#-NJGWB8z1ANLVtAdFa1s:-NJGWB8z1ANLVtAdFa1t:bg3ylsb) ([raw file](https://github.com/harness/harness-core/blob/07901e8fb46b0a3435748175932f7fcb24d8e333/950-delegate-tasks-beans/src/main/java/software/wings/beans/TaskType.java#L423)):*
> ```Java
>   ELASTIGROUP_PRE_FETCH_TASK_NG(TaskGroup.ELASTIGROUP, "Elastigroup Pre Fetch Task"),
> 
>   AWS_ASG_CANARY_DEPLOY_TASK_NG(TaskGroup.ASG, "AWS Asg Canary Deploy"),
> ```

Please don't move Deploy task, it was already there. It's dangerous to change order of already existing tasks. Please revert. you can add your task at the end.

**Comment by arvind-choudhary-h** [comment]
*[`400-rest/src/test/java/software/wings/beans/delegation/TaskTypeTest.java` line 419 at r6](https://reviewable.io/reviews/harness/harness-core/41429#-NJGWKlV8ainRC3TJp-_:-NJGWKlV8ainRC3TJp-a:bo4wvkf) ([raw file](https://github.com/harness/harness-core/blob/07901e8fb46b0a3435748175932f7fcb24d8e333/400-rest/src/test/java/software/wings/beans/delegation/TaskTypeTest.java#L419)):*
> ```Java
>     taskTypeOrdinalMapping.put(377, "ELASTIGROUP_SWAP_ROUTE_COMMAND_TASK_NG");
>     taskTypeOrdinalMapping.put(378, "ELASTIGROUP_ROLLBACK");
>     taskTypeOrdinalMapping.put(379, "ELASTIGROUP_PRE_FETCH_TASK_NG");
> ```

The existing task oridinals should be reverted back.

**Review by arvind-choudhary-h — COMMENTED** [review]
> Reviewed 13 of 15 files at r4, 6 of 6 files at r5, 7 of 7 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/41429)* status: 42 file(s) to review. 1 change requests. 0 of 2 approvals obtained.

**Review by arvind-choudhary-h — APPROVED** [review]
:lgtm:

> Reviewed 3 of 3 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/41429#-:-NJH-3ca34thjPzs7hr3:bnfp4nl)* status: 42 file(s) to review. 1 change requests. 1 of 2 approvals obtained.

**Comment by lovish1999** [comment]
trigger ti3

**Review by vitsafronovici — APPROVED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/41429#-:-NJL4u6B3RYXNKupYdBB:bnfp4nl)* status: 12 file(s) to review. 1 of 2 approvals obtained.

**Review by arvind-choudhary-h — APPROVED** [review]
> Reviewed 4 of 5 files at r8, 7 of 8 files at r9, 1 of 1 files at r10, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/41429)* status: :shipit: complete! ✓

**Pull request merged**